### PR TITLE
Update roadmap to reflect reality.

### PR DESCRIPTION
**Pull request by cpuguy83**
The roadmap is one of the most important ways that a new contributor may
get started on the codebase, as such it is important for it to reflect
the real effort that is currently happening.

This update just brings it up to date. There may be some other efforts
going on and I would encourage people to update the roadmap accordingly
as a separate effort.

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
SGTM :angel:

File: ROADMAP.md [hunk 1]
```diff
@@ -35,34 +35,64 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
 
 ## 1.1 Runtime improvements
 
-We introduced [`runC`](https://runc.io) as a standalone low-level tool for container
-execution in 2015, the first stage in spinning out parts of the Engine into standalone tools.
+Over time we have accumulated a lot of functionaility in the container runtime
```
Comment (thaJeztah) [inline ROADMAP.md]: s/functionaility/functionality/

File: ROADMAP.md [hunk 2]
```diff
@@ -35,34 +35,64 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+aspect of Moby while also growing in other areas. Much of the container runtime
+pieces are now duplicated work now available in other, lower level components
```
Comment (thaJeztah) [inline ROADMAP.md]: `now` is in this sentence twice; could remove one of them

File: ROADMAP.md [hunk 3]
```diff
@@ -35,34 +35,64 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+such as [containerd](https://containerd.io).
 
-As runC continued evolving, and the OCI specification along with it, we created
-[`containerd`](https://github.com/containerd/containerd), a daemon to control and monitor `runC`.
-In late 2016 this was relaunched as the `containerd` 1.0 track, aiming to provide a common runtime
-for the whole spectrum of container systems, including Kubernetes, with wide community support.
-This change meant that there was an increased scope for `containerd`, including image management
-and storage drivers.
+Moby currnetly only utilizes containerd for basic rutime state management, e.g. starting
```
Comment (thaJeztah) [inline ROADMAP.md]: s/currnetly/currently/

s/rutime/runtime/

File: ROADMAP.md [hunk 4]
```diff
@@ -35,34 +35,64 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
 
-## 1.2 Internal decoupling
+- #38043
+
+## 1.2 Image Builder
+
+Work is ongoing to integrate [BuildKit](https://github.com/moby/buildkit) into
+Moby and replace the "v0" build implementation.
+Buildkit offers better cache management, parallelizable build steps, and better
+extensibility while also keeping builds portable, a cheif tenent of Moby's builder.
```
Comment (thaJeztah) [inline ROADMAP.md]: s/cheif/chief/ (unless you were referring to [this urban dictionary use of that word](https://www.urbandictionary.com/define.php?term=Cheif) :joy:)

File: ROADMAP.md [hunk 5]
```diff
@@ -35,34 +35,64 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+
+Upon completion of this effort, users will have a builder that performs better while
+also being more extensible, enabling people to extend the Dockerfile format
+in their own way without being blocked by Moby maintainers, or even write a
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply ROADMAP.md]: Perhaps something like "a custom syntax" (it doesn't have to be a Dockerfile (like) syntax)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
Thanks for starting the work on this! I left some nits, and we'll definitely need more follow-ups, but it's a good start 👍

**Comment by thaJeztah** [comment]
As came up in the maintainers meeting, I think the big elephant in the room is also the API; how do we want to evolve the API? Split it to a separate repo (keep the definition/specs separate from implementation)? Head more towards a GRPC API, and keep the RESTful API as an optional compatibility layer?

We can keep that for a follow-up, but it's something we should discuss

**Review by iaoiui — CHANGES_REQUESTED** [review]

File: ROADMAP.md [hunk 6]
```diff
@@ -35,34 +35,63 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+Work is ongoing to integrate [BuildKit](https://github.com/moby/buildkit) into
+Moby and replace the "v0" build implementation.
+Buildkit offers better cache management, parallelizable build steps, and better
+extensibility while also keeping builds portable, a chief tenent of Moby's builder.
+
+Upon completion of this effort, users will have a builder that performs better while
+also being more extensible, enabling users to provide their own custom syntax
+which can be either Dockerfile-like or something completely different.
+
+See (buildpacks on buildkit)[https://github.com/tonistiigi/buildkit-pack] as an
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply ROADMAP.md]: Can you write what you'd like to see here and I'll update?

**Comment by cpuguy83** [comment]
@AkihiroSuda Added.

**Comment by thaJeztah** [comment]
ping @moby/moby-maintainers PTAL if things are missing that you'd like to have added

**Review by AkihiroSuda — APPROVED** [review]

**Review by vdemeester — APPROVED** [review]
LGTM :tiger:

File: ROADMAP.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpuguy83) [reply ROADMAP.md]: @tonistiigi WDYT?

[the remark this answers is not in the document]
Comment (tonistiigi) [reply ROADMAP.md]: New features for builder and Dockerfile should be implemented first in BuildKit backend using an external Dockerfile implementation from the container images. This allows everyone to test and evaluate the feature without upgrading their daemon. New features should go to experimental channel first and can be part of the `docker/dockerfile:experimental` image. From there they graduate to `docker/dockerfile:latest` and binary releases. The Dockerfile frontend source code is temporarily located at https://github.com/moby/buildkit/tree/master/frontend/dockerfile with separate new features defined with go build tags.

**Comment by AkihiroSuda** [comment]
Mergeable?

File: ROADMAP.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpuguy83) [reply ROADMAP.md]: Done

File: ROADMAP.md [hunk 7]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+Work is ongoing to integrate [BuildKit](https://github.com/moby/buildkit) into
+Moby and replace the "v0" build implementation. Buildkit offers better cache
+management, parallelizable build steps, and better extensibility while also
+keeping builds portable, a chief tenent of Moby's builder.
+
+Upon completion of this effort, users will have a builder that performs better
+while also being more extensible, enabling users to provide their own custom
+syntax which can be either Dockerfile-like or something completely different.
+
+See (buildpacks on buildkit)[https://github.com/tonistiigi/buildkit-pack] as an
```
Comment (thaJeztah) [inline ROADMAP.md]: The markdown for the link is flipped should be;

```suggestion
See [buildpacks on buildkit](https://github.com/tonistiigi/buildkit-pack) as an
```

File: ROADMAP.md [hunk 8]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+a signficant effort to refactor and even remove large parts of Moby's codebase.
 
-Moby will rely on a long-running `containerd` companion daemon for all container execution
-related operations. This could open the door in the future for Engine restarts without interrupting
-running containers. The switch over to containerd 1.0 is an important goal for the project, and
-will result in a significant simplification of the functions implemented in this repository.
+Tracking issues:
 
-## 1.2 Internal decoupling
+- #38043
```
Comment (thaJeztah) [inline ROADMAP.md]: Looks like GItHub won't render this as a link; can you make it a link? Perhaps add the title of the issue as well;

```suggestion
- [#38043](https://github.com/moby/moby/issues/38043) Proposal: containerd image integration
```

File: ROADMAP.md [hunk 9]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+their daemon. New features should go to experimental channel first and can be
+part of the `docker/dockerfile:experimental` image. From there they graduate to
+`docker/dockerfile:latest` and binary releases. The Dockerfile frontend source
+code is temporarily located at
+[https://github.com/moby/buildkit/tree/master/frontend/dockerfile](https://github.com/moby/buildkit/tree/master/frontend/dockerfile)
+with separate new features defined with go build tags.
+
+Tracking issues:
+
+- #32925
```
Comment (thaJeztah) [inline ROADMAP.md]: Same for this one;

```suggestion
- [#32925](https://github.com/moby/moby/issues/32925) discussion: builder future: buildki
```

File: ROADMAP.md [hunk 10]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+
+## 1.3 Rootless Mode
+
+Running the daemon requires elevated privileges for many tasks. We would like to
+support running the daemon as a nomral, unprivileged user without requiring suid
+binaries.
+
+Tracking issues:
+
+- #37375
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
- [#37375](https://github.com/moby/moby/issues/37375) Proposal: allow running `dockerd` as an unprivileged user (aka rootless mode)
```

File: ROADMAP.md [hunk 11]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+cover the full spectrum functionality and edge cases out there.
+
+Tests in the `integration-cli` folder should also be migrated into (both in
+location and style) the `integration` folder. These newer tests are simpler to
+run in isolation, simpler to read, simpler to write, and more fully exercise the
+API. Meanwhile tests of the docker CLI should generally live in docker/cli.
+
+Tracking issues:
+
+- #32866
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
- [#32866](https://github.com/moby/moby/issues/32866) Replace integration-cli suite with API test suite
```

File: ROADMAP.md [hunk 12]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+Moby and replace the "v0" build implementation. Buildkit offers better cache
+management, parallelizable build steps, and better extensibility while also
+keeping builds portable, a chief tenent of Moby's builder.
+
+Upon completion of this effort, users will have a builder that performs better
+while also being more extensible, enabling users to provide their own custom
+syntax which can be either Dockerfile-like or something completely different.
+
+See (buildpacks on buildkit)[https://github.com/tonistiigi/buildkit-pack] as an
+example this extensibility.
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
example of this extensibility.
```

File: ROADMAP.md [hunk 13]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+
+New features for the builder and Dockerfile should be implemented first in
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
New features for the builder and Dockerfile should be implemented first in the
```

File: ROADMAP.md [hunk 14]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+BuildKit backend using an external Dockerfile implementation from the container
+images. This allows everyone to test and evaluate the feature without upgrading
+their daemon. New features should go to experimental channel first and can be
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
their daemon. New features should go to the experimental channel first, and can be
```

File: ROADMAP.md [hunk 15]
```diff
@@ -35,34 +35,83 @@ issue, in the Slack channel, or in person at the Moby Summits that happen every
...
+with separate new features defined with go build tags.
+
+Tracking issues:
+
+- #32925
+
+## 1.3 Rootless Mode
+
+Running the daemon requires elevated privileges for many tasks. We would like to
+support running the daemon as a nomral, unprivileged user without requiring suid
```
Comment (thaJeztah) [inline ROADMAP.md]:
```suggestion
support running the daemon as a nomral, unprivileged user without requiring `suid`
```

**Review by thaJeztah — APPROVED** [review]
Left some minor suggestions and fixes, but LGTM after those are addressed

**Comment by AkihiroSuda** [comment]
@cpuguy83 could you address @thaJeztah 's comments?

**Comment by cpuguy83** [comment]
Updated.

**Comment by olljanat** [comment]
> As came up in the maintainers meeting, I think the big elephant in the room is also the API; how do we want to evolve the API? Split it to a separate repo (keep the definition/specs separate from implementation)?

Based on my experience so far I think that it would be good idea to move at least ```github.com/docker/docker/api/types``` to own repository. Now some changes like this one https://github.com/moby/moby/issues/25885#issuecomment-447657852 needs quite boring PR loop between repositories.

> We can keep that for a follow-up, but it's something we should discuss

Should we create separate issue for that discussion?

**Comment by cpuguy83** [comment]
I don't think it belongs in the roadmap, for sure.

**Comment by olljanat** [comment]
Derek add label: rebuild/janky

**Comment by thaJeztah** [comment]
There's this issue as well, so we can use that for discussing; https://github.com/moby/moby/issues/32873

**Review by thaJeztah — APPROVED** [review]
whoops, see I didn't LGTM since you updated 😊 😅

**Comment by thaJeztah** [comment]
Let's merge this; I see there's enough LGTM's, and we can update in follow-ups if desired 👍

**Pull request merged**